substrate/primitives/transaction-storage-proof/src/lib.rs

```diff
@@ -207,33 +219,39 @@ pub mod registration {
 				for (index, chunk) in chunks.enumerate() {
 					let index = encode_index(index as u32);
 					trie.insert(&index, &chunk).map_err(|e| Error::Application(Box::new(e)))?;
-					if chunk_index == target_chunk_index {
-						target_chunk = Some(chunk);
-						target_chunk_key = index;
+					if chunk_index == selected_chunk_index {
+						selected_chunk_and_key = Some((chunk, index));
 					}
 					chunk_index += 1;
 				}
 				trie.commit();
 			}
-			if target_chunk.is_some() && target_root == Default::default() {
-				target_root = transaction_root;
-				chunk_proof = sp_trie::generate_trie_proof::<TrieLayout, _, _, _>(
+			if let Some((target_chunk, target_chunk_key)) = selected_chunk_and_key {
+				let chunk_proof = sp_trie::generate_trie_proof::<TrieLayout, _, _, _>(
 					&db,
 					transaction_root,
-					&[target_chunk_key.clone()],
+					&[target_chunk_key],
 				)
 				.map_err(|e| Error::Application(Box::new(e)))?;
+
+				// We found the chunk and computed the proof root for the entire transaction,
+				// so there is no need to waste time calculating the subsequent transactions.
+				return Ok(Some(TransactionStorageProof {
+					proof: chunk_proof,
+					chunk: target_chunk,
+				}));
 			}
 		}
 
-		Ok(TransactionStorageProof { proof: chunk_proof, chunk: target_chunk.unwrap() })
+		debug_assert!(false, "No chunk matched the target_chunk_index; logic error");
```

Comment:
Probably returning an error would be better as I assume that the block would fail to produce any way?